src/util/rocksdb.rs

```diff
@@ -226,12 +235,176 @@ impl SliceTransform for NoopSliceTransform {
+const PROP_NUM_ROWS: &'static str = "tikv.num_rows";
+const PROP_NUM_PUTS: &'static str = "tikv.num_puts";
+const PROP_NUM_DELETES: &'static str = "tikv.num_delets";
+const PROP_NUM_VERSIONS: &'static str = "tikv.num_versions";
+
+#[derive(Clone, Debug, Default)]
+pub struct UserProperties {
+    pub min_ts: u64,
+    pub max_ts: u64,
+    pub num_rows: u64,
+    pub num_puts: u64,
+    pub num_deletes: u64,
+    pub num_versions: u64,
+}
+
+impl UserProperties {
+    pub fn new() -> UserProperties {
+        UserProperties {
+            min_ts: u64::MAX,
+            max_ts: u64::MIN,
+            num_rows: 0,
+            num_puts: 0,
+            num_deletes: 0,
+            num_versions: 0,
+        }
+    }
+
+    pub fn add(&mut self, other: &UserProperties) {
+        self.min_ts = cmp::min(self.min_ts, other.min_ts);
+        self.max_ts = cmp::max(self.max_ts, other.max_ts);
+        self.num_rows += other.num_rows;
+        self.num_puts += other.num_puts;
+        self.num_deletes += other.num_deletes;
+        self.num_versions += other.num_versions;
+    }
+
+    pub fn encode(&self) -> HashMap<Vec<u8>, Vec<u8>> {
+        let items = [(PROP_MIN_TS, self.min_ts),
+                     (PROP_MAX_TS, self.max_ts),
+                     (PROP_NUM_ROWS, self.num_rows),
+                     (PROP_NUM_PUTS, self.num_puts),
+                     (PROP_NUM_DELETES, self.num_deletes),
+                     (PROP_NUM_VERSIONS, self.num_versions)];
+        items.iter()
+            .map(|&(k, v)| {
+                let mut buf = Vec::with_capacity(8);
+                buf.encode_u64(v).unwrap();
+                (k.as_bytes().to_owned(), buf)
+            })
+            .collect()
+    }
+
+    pub fn decode<T: DecodeU64>(props: &T) -> Result<UserProperties, codec::Error> {
+        let mut res = UserProperties::new();
+        res.min_ts = try!(props.decode_u64(PROP_MIN_TS));
+        res.max_ts = try!(props.decode_u64(PROP_MAX_TS));
+        res.num_rows = try!(props.decode_u64(PROP_NUM_ROWS));
+        res.num_puts = try!(props.decode_u64(PROP_NUM_PUTS));
+        res.num_deletes = try!(props.decode_u64(PROP_NUM_DELETES));
+        res.num_versions = try!(props.decode_u64(PROP_NUM_VERSIONS));
+        Ok(res)
+    }
+}
+
+pub struct UserPropertiesCollector {
+    props: UserProperties,
+    last_key: Vec<u8>,
+}
+
+impl UserPropertiesCollector {
+    fn new() -> UserPropertiesCollector {
+        UserPropertiesCollector {
+            props: UserProperties::new(),
+            last_key: Vec::new(),
+        }
+    }
+}
+
+impl TablePropertiesCollector for UserPropertiesCollector {
+    fn add(&mut self, key: &[u8], value: &[u8], entry_type: DBEntryType, _: u64, _: u64) {
+        if !keys::validate_data_key(key) {
+            return;
+        }
+
+        let (k, ts) = match types::split_encoded_key_on_ts(key) {
+            Ok((k, ts)) => (k, ts),
+            Err(_) => return,   // Ignore error
```

Comment:
`UserProperties` aims to be a common type returned by the snapshot interface.
It can be split into multiple parts when it's too large, but we have only a few properties right now.